system/jlib/jsecrets.cpp

```diff
@@ -1364,106 +1311,233 @@ jlib_decl bool containsEmbeddedKey(const char *certificate)
+        ca->setProp("pem", value.str());
+    }
+}
 
-    validateSecretName(name);
 
-    CriticalBlock block(mtlsInfoCacheCS);
-    Owned<IPropertyTree> info = mtlsInfoCache->getPropTree(name);
-    if (info)
-        return info.getClear();
+//---------------------------------------------------------------------------------------------------------------------
 
-    StringBuffer filepath;
-    StringBuffer secretpath;
+class CIssuerConfig final : public CSyncedCertificateBase
+{
+public:
+    CIssuerConfig(const char *_issuer, const char * _trustedPeers, bool _isClientConnection, bool _acceptSelfSigned, bool _addCACert, bool _disableMTLS)
+    : CSyncedCertificateBase(_issuer), trustedPeers(_trustedPeers), isClientConnection(_isClientConnection), acceptSelfSigned(_acceptSelfSigned), addCACert(_addCACert), disableMTLS(_disableMTLS)
+    {
+        secret.setown(resolveSecret("certificates", issuer, nullptr, nullptr));
+        createConfig();
+    }
 
-    buildSecretPath(secretpath, "certificates", name);
+    virtual void updateConfigFromSecret(const IPropertyTree * secretInfo) const override;
 
-    filepath.set(secretpath).append("tls.crt");
-    if (!checkFileExists(filepath))
-        return nullptr;
+protected:
+    StringAttr trustedPeers;
+    bool isClientConnection; // required in constructor
+    bool acceptSelfSigned; // required in constructor
+    bool addCACert; // required in constructor
+    bool disableMTLS;
+};
 
-    info.set(mtlsInfoCache->setPropTree(name));
-    info->setProp("@issuer", name);
-    info->setProp("certificate", filepath.str());
-    filepath.set(secretpath).append("tls.key");
-    if (checkFileExists(filepath))
-        info->setProp("privatekey", filepath.str());
-    IPropertyTree *verify = ensurePTree(info, "verify");
-    if (verify)
-    {
-        filepath.set(secretpath).append("ca.crt");
-        if (checkFileExists(filepath))
-        {
-            IPropertyTree *ca = ensurePTree(verify, "ca_certificates");
-            if (ca)
-                ca->setProp("@path", filepath.str());
-        }
-        //For now only the "public" issuer implies client certificates are not required
-        verify->setPropBool("@enable", !strieq(name, "public"));
-        verify->setPropBool("@address_match", false);
-        verify->setPropBool("@accept_selfsigned", false);
+
+void CIssuerConfig::updateConfigFromSecret(const IPropertyTree * secretInfo) const
+{
+    if (!isClientConnection || !strieq(issuer, "public"))
+        updateCertificateFromSecret(secretInfo);
+
+    if (!isClientConnection || addCACert) // same condition as before, but is it correct?
+        updateCertificateAuthorityFromSecret(secretInfo);
+
+    IPropertyTree *verify = config->queryPropTree("verify");
+    //For now only the "public" issuer implies client certificates are not required
+    verify->setPropBool("@enable", !disableMTLS && (isClientConnection || !strieq(issuer, "public")));
```

Comment:
is 'verify' guaranteed to exist by this state? does it need to be null checked.

Reply:
yes - I will add an assert for safety

Reply:
yes meaning it is guaranteed to exist